### PR TITLE
feat(streaming): introduce cache policy for hash join

### PR DESCRIPTION
**Pull request by yuhao-su**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Previously our hash join executor only populate cache on reading a join key. Now we allow users to specify whether to populate cache on read or on both read an write.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

To populate cache on read (default setting) use the following statement:
```
SET RW_STREAMING_HASH_JOIN_POLICY TO OnRead;
```

To read cache on both read and write use the following statement:
```
SET RW_STREAMING_HASH_JOIN_POLICY TO OnReadWrite;
```
### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 1]
```diff
@@ -113,6 +113,29 @@ pub type HashValueType = JoinEntryState;
...
+
+    use super::JoinCachePolicyPrimitive;
+    use crate::executor::error::StreamExecutorResult;
+
+    /// Populate the cache when read a join key
+    pub const OnRead: JoinCachePolicyPrimitive = 0;
+    /// Populate the cache when read or write a join key
+    pub const OnReadWrite: JoinCachePolicyPrimitive = 1;
+
+    pub fn from_prost(policy_prost: &str) -> StreamExecutorResult<JoinCachePolicyPrimitive> {
```
Comment (BugenZhao) [inline src/stream/src/executor/managed_state/join/mod.rs:130]: How about defining as a protobuf enum?

**Comment by fuyufjh** [comment]
1. IIUC, your goal is to save a write, but actually since there could be more than 1 rows in a join entry, either `OnRead` or `OnReadWrite` needs to use a read to populate a join entry entirely, so I didn't see any difference. Even worse, if the join is between a fact table and a dimension table, populating cache on the fact table side is likely to be useless, because there are very few changes on the dimension table side.
2. I don't like the idea of configurable `RW_STREAMING_HASH_JOIN_POLICY` because users must never understand these internal details and *nobody* will use this config in the future. Please just select a generally good policy and remove others.

**Comment by yuhao-su** [comment]
Preliminary benching (tpch q12) shows both policy produced similar result. However we are currently facing a compactor OOM issue that causes performance deterioration within 5 min. I will bench it later after #3921 solved.

**Comment by lmatz** [comment]
Will this PR be part of #5395 or #5395 will make this PR obsolete?

**Comment by yuhao-su** [comment]
> Will this PR be part of #5395 or #5395 will make this PR obsolete?

No. They are not related.

**Comment by TennyZhuang** [comment]
Close due to outdated and controversies.

**Pull request closed without merging**